rust-runtime/aws-smithy-runtime/src/client/runtime_plugin/client_rate_limiter.rs

```diff
@@ -0,0 +1,597 @@
+        Self::builder()
+            .time_of_last_throttle(get_unix_timestamp(cfg))
+            .previous_time_bucket(get_unix_timestamp(cfg).floor())
+            .build()
+    }
+
+    fn builder() -> Builder {
+        Builder::new()
+    }
+
+    /// If this function returns `Ok(())`, you're OK to send a request. If it returns an error,
+    /// then you should not send a request; You've sent quite enough already.
+    pub(crate) fn acquire_permission_to_send_a_request(
+        &mut self,
+        cfg: &ConfigBag,
+        amount: f64,
+    ) -> Result<(), BoxError> {
+        if !self.enable_throttling {
+            // return early if we haven't encountered a throttling error yet
+            return Ok(());
+        }
+
+        let timestamp = get_unix_timestamp(cfg);
+        self.refill(timestamp);
+
+        if self.current_bucket_capacity < amount {
+            Err(BoxError::from("the client rate limiter is out of tokens"))
+        } else {
+            self.current_bucket_capacity -= amount;
+            Ok(())
+        }
+    }
+
+    pub(crate) fn update_rate_limiter(&mut self, cfg: &ConfigBag, is_throttling_error: bool) {
+        let timestamp = get_unix_timestamp(cfg);
+        self.update_tokens_retrieved_per_second(timestamp);
+
+        if is_throttling_error {
+            let rate_to_use = if self.enable_throttling {
+                f64::min(self.tokens_retrieved_per_second, self.token_refill_rate)
+            } else {
+                self.tokens_retrieved_per_second
+            };
+
+            // The fill_rate is from the token bucket
+            self.tokens_retrieved_per_second_at_time_of_last_throttle = rate_to_use;
+            self.calculate_time_window();
+            self.time_of_last_throttle = timestamp;
+            self.calculated_rate = cubic_throttle(rate_to_use);
+            self.enable_token_bucket();
+        } else {
+            self.calculate_time_window();
+            self.calculated_rate = self.cubic_success(timestamp);
+        }
+
+        let new_rate = f64::min(self.calculated_rate, 2.0 * self.tokens_retrieved_per_second);
+        self.update_bucket_refill_rate(timestamp, new_rate);
+    }
+
+    fn refill(&mut self, timestamp: f64) {
+        if let Some(last_timestamp) = self.time_of_last_refill {
+            let fill_amount = (timestamp - last_timestamp) * self.token_refill_rate;
+            self.current_bucket_capacity = f64::min(
+                self.maximum_bucket_capacity,
+                self.current_bucket_capacity + fill_amount,
+            );
+        }
+        self.time_of_last_refill = Some(timestamp);
+    }
+
+    fn update_bucket_refill_rate(&mut self, timestamp: f64, new_fill_rate: f64) {
+        // Refill based on our current rate before we update to the new fill rate.
+        self.refill(timestamp);
+
+        self.token_refill_rate = f64::max(new_fill_rate, MIN_FILL_RATE);
+        self.maximum_bucket_capacity = f64::max(new_fill_rate, MIN_CAPACITY);
```

Comment:
To better understand, can fill rate be assignable to a bucket capacity?

Reply:
I don't actually understand this part too well. I took the implementation from the Go v2 SDK.